rrs/models.py

```diff
@@ -0,0 +1,566 @@
+
+class RecipeUpstream(models.Model):
+    RECIPE_UPSTREAM_STATUS_CHOICES = (
+        ('A', 'All'),
+        ('N', 'Not updated'),
+        ('C', 'Can\'t be updated'),
+        ('Y', 'Up-to-date'),
+        ('D', 'Downgrade'),
+        ('U', 'Unknown'),
+    )
+    RECIPE_UPSTREAM_STATUS_CHOICES_DICT = dict(RECIPE_UPSTREAM_STATUS_CHOICES)
+
+    RECIPE_UPSTREAM_TYPE_CHOICES = (
+        ('A', 'Automatic'),
+        ('M', 'Manual'),
+    )
+    RECIPE_UPSTREAM_TYPE_CHOICES_DICT = dict(RECIPE_UPSTREAM_TYPE_CHOICES)
+
+    recipesymbol = models.ForeignKey(RecipeSymbol)
+    history = models.ForeignKey(RecipeUpstreamHistory)
+    version = models.CharField(max_length=100, blank=True)
+    type = models.CharField(max_length=1, choices=RECIPE_UPSTREAM_TYPE_CHOICES, blank=True, db_index=True)
+    status =  models.CharField(max_length=1, choices=RECIPE_UPSTREAM_STATUS_CHOICES, blank=True, db_index=True)
+    no_update_reason = models.CharField(max_length=255, blank=True, db_index=True)
+    date = models.DateTimeField(db_index=True)
+
+    @staticmethod
+    def get_all_recipes(history):
+        qry = RecipeUpstream.objects.filter(history = history)
+        return qry
+
+    @staticmethod
+    def get_recipes_not_updated(history):
+        qry = RecipeUpstream.objects.filter(history = history, status = 'N',
+                no_update_reason = '').order_by('pn')
+        return qry
+
+    @staticmethod
+    def get_recipes_cant_be_updated(history):
+        qry = RecipeUpstream.objects.filter(history = history, status = 'N') \
+                .exclude(no_update_reason = '').order_by('pn')
+        return qry
+
+    @staticmethod
+    def get_recipes_up_to_date(history):
+        qry = RecipeUpstream.objects.filter(history = history, status = 'Y' \
+                ).order_by('pn')
+        return qry
+
+    @staticmethod
+    def get_recipes_unknown(history):
+        qry = RecipeUpstream.objects.filter(history = history,
+                status__in = ['U', 'D']).order_by('pn')
+        return qry
+
+    @staticmethod
+    def get_by_recipe_and_history(recipesymbol, history):
+        ru = RecipeUpstream.objects.filter(recipesymbol=recipesymbol, history=history)
+        return ru[0] if ru else None
+
+    def needs_upgrade(self):
+        if self.status == 'N':
+            return True
+        else:
+            return False
+
+    def __str__(self):
+        return '%s: (%s, %s, %s)' % (self.recipesymbol.pn, self.status,
+                self.version, self.date)
+
+class RecipeDistro(models.Model):
+    recipe = models.ForeignKey(Recipe)
+    distro = models.CharField(max_length=100, blank=True)
+    alias = models.CharField(max_length=100, blank=True)
+
+    def __str__(self):
+        return '%s: %s' % (self.recipe.pn, self.distro)
+
+    @staticmethod
+    def get_distros_by_recipe(recipe):
+        recipe_distros = []
+
+        query = RecipeDistro.objects.filter(recipe = recipe).order_by('distro')
+        for q in query:
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To create a `list`, try to use `list` comprehension instead of a loop. List comprehension is the preferred way to make a list using Python, and it&#39;s simpler and easier to understand than using a loop.

[Learn more](https://docs.python.org/3/tutorial/datastructures.html#list-comprehensions)